sim/network.go

```diff
@@ -108,49 +120,49 @@ func (n *Network) BroadcastSynchronous(sender gpbft.ActorID, msg gpbft.GMessage)
 	}
 }
 
-// Returns whether there are any more messages to process.
+// Tick disseminates one message among participants and returns whether there are
+// any more messages to process.
 func (n *Network) Tick(adv *adversary.Adversary) (bool, error) {
-	// Find first message the adversary will allow.
-	i := 0
-	if adv != nil && !n.globalStabilisationElapsed {
-		for ; i < len(n.queue); i++ {
-			msg := n.queue[i]
-			gmsg, ok := msg.payload.(gpbft.GMessage)
-			if !ok || adv.AllowMessage(msg.source, msg.dest, gmsg) {
-				break
-			}
-		}
-		// If adversary blocks everything, assume GST has passed.
-		if i == len(n.queue) {
-			n.Log("GST elapsed")
-			n.globalStabilisationElapsed = true
-			i = 0
-		}
-	}
+	msg := n.queue.Remove()
+	n.clock = msg.deliverAt
 
-	msg := n.queue.Remove(i)
-	if msg.deliverAt.After(n.clock) {
-		n.clock = msg.deliverAt
+	receiver, found := n.participants[msg.dest]
+	if !found {
+		return false, fmt.Errorf("message destined to unknown participant ID: %d", msg.dest)
 	}
-	payloadStr, ok := msg.payload.(string)
-	receiver := n.participants[msg.dest]
-	if ok && strings.HasPrefix(payloadStr, "ALARM") {
-		n.log(TraceRecvd, "P%d %s", msg.source, payloadStr)
+	switch payload := msg.payload.(type) {
+	case string:
+		if payload != "ALARM" {
+			return false, fmt.Errorf("unknwon string message payload: %s", payload)
+		}
+		n.log(TraceRecvd, "P%d %s", msg.source, payload)
 		if err := receiver.ReceiveAlarm(); err != nil {
-			return false, fmt.Errorf("failed receiving alarm: %w", err)
+			return false, fmt.Errorf("failed to deliver alarm from %d to %d: %w", msg.source, msg.dest, err)
+		}
+	case gpbft.GMessage:
+		// If GST has not elapsed, check if adversary allows the propagation of message.
+		if adv != nil && !n.globalStabilisationElapsed {
+			if n.hasGlobalStabilizationTimeElapsed() {
```

Comment:
This is a bit confusing to read. We should just remove the boolean field if this can be calculated instead.

Reply:
Ok makes sense, thanks.